src/lib/storage/defragment_reference_table.cpp

```diff
@@ -0,0 +1,125 @@
+#include "defragment_reference_table.hpp"
+
+#include "base_column.hpp"
+#include "chunk.hpp"
+#include "reference_column.hpp"
+#include "table.hpp"
+
+namespace opossum {
+
+std::shared_ptr<Table> defragment_reference_table(const std::shared_ptr<const Table>& reference_table,
+                                                  const ChunkOffset min_chunk_size, const ChunkOffset max_chunk_size) {
+  DebugAssert(reference_table->type() == TableType::References, "Can't handle non-reference tables");
+
+  const auto output_table = std::make_shared<Table>(reference_table->column_definitions(), TableType::References,
+                                                    reference_table->max_chunk_size());
+
+  for (auto chunk_id_begin = ChunkID{0}; chunk_id_begin < reference_table->chunk_count();) {
+    const auto chunk_begin = reference_table->get_chunk(chunk_id_begin);
+
+    /**
+     * I. Figure out which consecutive Chunks to merge into one. Stored as range [chunk_id_begin, chunk_id_end)
+     *
+     * For a Chunk to be included in the range
+     *      - its columns must reference to the same table and ColumnID as the other Columns with the same ColumnID.
+     *      - it must have <= rows than `min_chunk_size`
+     *      - adding it to the range mustn't result in a merged chunk bigger than `max_chunk_size`
+     */
+    auto accumulated_row_count = chunk_begin->size();              // Number of rows in the merged Chunk
+    auto chunk_id_end = static_cast<ChunkID>(chunk_id_begin + 1);  // ChunkID after the last Chunk to merge
+    for (; chunk_id_end < reference_table->chunk_count(); ++chunk_id_end) {
+      const auto chunk_end = reference_table->get_chunk(chunk_id_end);
+
+      // If chunk_end is too big itself or merging it with the Chunks above would yield in a Chunk too big, don't
+      // include it in the merging
+      if (chunk_end->size() > min_chunk_size || accumulated_row_count + chunk_end->size() > max_chunk_size) {
+        break;
+      }
+
+      // If the Chunk references different tables/columns than the one above, don't merge it.
+      auto column_layout_matches = true;
+      for (auto column_id = ColumnID{0}; column_id < reference_table->column_count(); ++column_id) {
+        const auto reference_column_a =
+            std::static_pointer_cast<const ReferenceColumn>(chunk_begin->get_column(column_id));
+        const auto reference_column_b =
+            std::static_pointer_cast<const ReferenceColumn>(chunk_end->get_column(column_id));
+
+        if (reference_column_a->referenced_table() != reference_column_b->referenced_table() ||
+            reference_column_a->referenced_column_id() != reference_column_b->referenced_column_id()) {
+          column_layout_matches = false;
+          break;
```

Comment:
breaks in nested for loops are nasty. Please add a comment on what the breaks mean. (also, `break n` is one of the very few good things in PHP...)